contracts/account/manager/src/commands.rs

```diff
@@ -291,72 +308,119 @@ pub fn upgrade_modules(
+            old_module_addr,
+            requested_module.info,
+            code_id,
+        )?,
+        ModuleReference::AccountBase(code_id) | ModuleReference::Standalone(code_id) => {
+            vec![build_module_migrate_msg(
                 old_module_addr,
                 code_id,
-                migrate_msg.unwrap_or_else(|| to_binary(&Empty {}).unwrap()),
-            ));
+                migrate_msg.unwrap(),
+            )]
         }
-        ModuleReference::AccountBase(code_id) | ModuleReference::Standalone(code_id) => msgs.push(
-            get_migrate_msg(old_module_addr, code_id, migrate_msg.unwrap()),
-        ),
+
         _ => return Err(ManagerError::NotUpgradeable(module_info)),
     };
+    msgs.extend(migrate_msgs);
+    Ok(())
+}
+
+/// Handle API module migration and return the migration messages
+fn handle_api_migration(
+    mut deps: DepsMut,
+    module_info: ModuleInfo,
+    old_api_addr: Addr,
+    new_api_addr: Addr,
+) -> ManagerResult<Vec<CosmosMsg>> {
+    let module_id = module_info.id();
+    versioning::assert_migrate_requirements(
+        deps.as_ref(),
+        &module_id,
+        module_info.version.try_into()?,
+    )?;
+    let old_deps = versioning::load_module_dependencies(deps.as_ref(), &module_id)?;
+    // Update the address of the api internally
+    update_module_addresses(
+        deps.branch(),
+        Some(vec![(module_id.clone(), new_api_addr.to_string())]),
+        None,
+    )?;
+
+    add_module_upgrade_to_context(deps.storage, &module_id, old_deps)?;
+
+    replace_api(deps, new_api_addr, old_api_addr)
+}
+
+/// Handle app module migration and return the migration messages
+fn handle_app_migration(
+    deps: DepsMut,
+    migrate_msg: Option<Binary>,
+    old_module_addr: Addr,
+    module_info: ModuleInfo,
+    code_id: u64,
+) -> ManagerResult<Vec<CosmosMsg>> {
+    let module_id = module_info.id();
+    versioning::assert_migrate_requirements(
+        deps.as_ref(),
+        &module_id,
+        module_info.version.try_into()?,
+    )?;
+    let old_deps = versioning::load_module_dependencies(deps.as_ref(), &module_id)?;
+
+    // Add module upgrade to reply context
+    add_module_upgrade_to_context(deps.storage, &module_id, old_deps)?;
+
+    Ok(vec![build_module_migrate_msg(
+        old_module_addr,
+        code_id,
+        migrate_msg.unwrap_or_else(|| to_binary(&Empty {}).unwrap()),
+    )])
+}
+
+/// Add the module upgrade to the migration context and check for duplicates
+fn add_module_upgrade_to_context(
+    storage: &mut dyn Storage,
+    module_id: &str,
+    module_deps: Vec<Dependency>,
+) -> Result<(), ManagerError> {
+    // Add module upgrade to reply context
+    let update_context = |mut upgraded_modules: Vec<(String, Vec<Dependency>)>| -> StdResult<Vec<(String, Vec<Dependency>)>> {
+        if upgraded_modules.iter().any(|(id, _)| id == module_id) {
+            Err(StdError::generic_err("Module already upgraded.".to_string()))
```

Comment:
Is this check not better performed before any messages/state is changed? Right when we parse the message we can check if there are duplicate migration requests. Should be cheaper gas wise.

Reply:
Nice catch!